repo-client/src/main/kotlin/io/lionweb/repoclient/LionWebClient.kt

```diff
@@ -0,0 +1,734 @@
+                var ids = rootIds.toString()
+                if (ids.length > 100) {
+                    ids = ids.substring(0, 100)
+                }
+                debugFile("retrieved-$ids.json") { data }
+
+                return processChunkResponse(data) {
+                    val js = jsonSerialization
+                    js.unavailableParentPolicy =
+                        if (withProxyParent) {
+                            UnavailableNodePolicy.PROXY_NODES
+                        } else {
+                            UnavailableNodePolicy.NULL_REFERENCES
+                        }
+                    js.unavailableReferenceTargetPolicy = UnavailableNodePolicy.PROXY_NODES
+                    js.unavailableChildrenPolicy =
+                        when (retrievalMode) {
+                            RetrievalMode.ENTIRE_SUBTREE -> UnavailableNodePolicy.THROW_ERROR
+                            RetrievalMode.SINGLE_NODE -> UnavailableNodePolicy.PROXY_NODES
+                        }
+                    val nodes = js.deserializeToNodes(it)
+                    rootIds.map { rootId ->
+                        nodes.find { node -> node.id == rootId } ?: throw IllegalArgumentException(
+                            "When requesting a subtree with rootId=$rootId we got back an answer without such ID. " +
+                                "IDs we got back: ${nodes.map { node -> node.id }.joinToString(", ")}",
+                        )
+                    }
+                }
+            } else {
+                throw RuntimeException(
+                    "Something went wrong while querying $url: http code ${response.code}, body: ${response.body?.string()}",
+                )
+            }
+        }
+    }
+
+    fun getAncestorsId(nodeID: String): List<String> {
+        val result = mutableListOf<String>()
+        var currentNodeID: String? = nodeID
+        while (currentNodeID != null) {
+            currentNodeID = getParentId(currentNodeID)
+            if (currentNodeID != null) {
+                result.add(currentNodeID)
+            }
+        }
+        return result
+    }
+
+    fun <T : Node> retrieveAncestor(
+        node: Node,
+        ancestorClass: KClass<*>,
+        retrievalMode: RetrievalMode = RetrievalMode.ENTIRE_SUBTREE,
+    ): T? {
+        if (node.parent == null) {
+            return null
+        } else {
+            var parent = node.parent
+            if (node.parent is ProxyNode) {
+                parent = retrieve(node.parent.id!!, withProxyParent = true, retrievalMode = RetrievalMode.SINGLE_NODE)
+            }
+            return if (ancestorClass.isInstance(parent)) {
+                if (retrievalMode == RetrievalMode.SINGLE_NODE) {
+                    parent as T
+                } else {
+                    retrieve(node.parent.id!!, withProxyParent = true, retrievalMode = RetrievalMode.ENTIRE_SUBTREE) as T
```

Comment:
Why `RetrievalModel.ENTIRE_SUBTREE`, as we are looking only at the parent?

Reply:
Bacause here we have identified the ancestor we are interested into, and we can now retrieve the entire subtree of the ancestor, so that it can be properly navigated. Otherwise all children would be proxies and so the user would need to do further requests

Reply:
I understand, but doesn't this mean that for each parent upwards the full subtree will be retrieved. That might result in getting quite a lot of the same nodes with each upwards parent move.

Reply:
No, because it does not happen for each parent but only when we identified already the right parent (see `if (ancestorClass.isInstance(parent)) {...`.

In practice we keep asking for parents in shallow mode (i.e., `retrievalMode == RetrievalMode.SINGLE_NODE`). Once we found the node of the expected type, if the user specified they want the shallow version we just return it, if instead the user specified they want the deep version we do another retrieve but this time we specify  `retrievalMode = RetrievalMode.ENTIRE_SUBTRE`. This means this particular ancestor will be retrieved twice: one in the shallow mode (`parent = retrieve(node.parent.id!!, withProxyParent = true, retrievalMode = RetrievalMode.SINGLE_NODE)` and then, when we know we have a match, again in deep mode (`retrieve(node.parent.id!!, withProxyParent = true, retrievalMode = RetrievalMode.ENTIRE_SUBTREE) as T`)

Reply:
You are right, I read the code incorrectly.